src/linear_algebra/general_linear_group.lean

```diff
@@ -0,0 +1,186 @@
+
+/--The `GL n R` and `general_linear_group R n` groups are multiplicatively equivalent-/
+def to_lin : (GL n R) ≃* (linear_map.general_linear_group R (n → R)) :=
+units.map_equiv to_lin_alg_equiv'.to_mul_equiv
+
+/--Given a matrix with invertible determinant we get an element of `GL n R`-/
+def mk' (A : matrix n n R) (h : invertible (matrix.det A)) : GL n R :=
+unit_of_det_invertible A
+
+/--Given a matrix with unit determinant we get an element of `GL n R`-/
+noncomputable def mk'' (A : matrix n n R) (h : is_unit (matrix.det A)) : GL n R :=
+nonsing_inv_unit A h
+
+instance coe_fun : has_coe_to_fun (GL n R) :=
+{ F   := λ _, n → n → R,
+  coe := λ A, A.val }
+
+lemma ext_iff (A B : GL n R) : A = B ↔ (∀ i j, (A : matrix n n R) i j = (B : matrix n n R) i j) :=
+units.ext_iff.trans matrix.ext_iff.symm
+
+/-- Not marked `@[ext]` as the `ext` tactic already solves this. -/
+lemma ext ⦃A B : GL n R⦄ (h : ∀ i j, (A : matrix n n R) i j = (B : matrix n n R) i j) :
+  A = B :=
+units.ext $ matrix.ext h
+
+section coe_lemmas
+
+variables (A B : GL n R)
+
+@[simp] lemma coe_fn_eq_coe : ⇑A = (↑A : matrix n n R) := rfl
+
+@[simp] lemma coe_mul : ↑(A * B) = (↑A : matrix n n R) ⬝ (↑B : matrix n n R) := rfl
+
+@[simp] lemma coe_one : ↑(1 : GL n R) = (1 : matrix n n R) := rfl
+
+@[simp] lemma coe_inv : ↑(A⁻¹) = (↑A : matrix n n R)⁻¹ :=
+begin
+  letI := A.invertible,
+  exact inv_eq_nonsing_inv_of_invertible (↑A : matrix n n R),
+end
+
+end coe_lemmas
+
+end general_linear_group
+
+namespace special_linear_group
+
+variables {n : Type u} [decidable_eq n] [fintype n] {R : Type v} [comm_ring R]
+
+instance has_coe_to_general_linear_group : has_coe (special_linear_group n R) (GL n R) :=
+⟨λ A, ⟨↑A, ↑(A⁻¹), congr_arg coe (mul_right_inv A), congr_arg coe (mul_left_inv A)⟩⟩
+
+end special_linear_group
+
+section
+
+variables {n : Type u} {R : Type v} [decidable_eq n] [fintype n] [linear_ordered_comm_ring R ]
+
+section
+variables (n R)
+
+/-- This is the subgroup of `nxn` matrices with entries over a
+linear ordered ring and positive determinant. -/
+def GL_pos : subgroup (GL n R) :=
+units.pos_subgroup.comap general_linear_group.det
+
+end
+
+@[simp] lemma mem_GL_pos (A : GL n R) : A ∈ GL_pos n R ↔ 0 < (A.det : R) := iff.rfl
+
+end
+
+section has_neg
+
+variables {n : Type u} {R : Type v} [decidable_eq n] [fintype n] [linear_ordered_comm_ring R ]
+[fact (even (fintype.card n))]
+
+/-- Formal operation of negation on general linear group on even cardinality `n` given by negating
+each element. -/
+instance : has_neg (GL_pos n R) :=
+⟨λ g,
+  ⟨- g, by {simp only [mem_GL_pos, general_linear_group.coe_det_apply, units.coe_neg],
+  have := det_smul g (-1),
+  simp only [general_linear_group.coe_fn_eq_coe, one_smul, coe_fn_coe_base, neg_smul] at this,
+  rw this,
+  simp [nat.neg_one_pow_of_even (fact.out (even (fintype.card n)))],
+  have gdet:=g.property,
+  simp only [mem_GL_pos, general_linear_group.coe_det_apply, subtype.val_eq_coe] at gdet,
+  exact gdet,}⟩⟩
```

Comment:
```suggestion
  ⟨- g,
  begin
    simp only [mem_GL_pos, general_linear_group.coe_det_apply, units.coe_neg],
    have := det_smul g (-1),
    simp only [general_linear_group.coe_fn_eq_coe, one_smul, coe_fn_coe_base, neg_smul] at this,
    rw this,
    simp [nat.neg_one_pow_of_even (fact.out (even (fintype.card n)))],
    have gdet := g.property,
    simp only [mem_GL_pos, general_linear_group.coe_det_apply, subtype.val_eq_coe] at gdet,
    exact gdet,
  end⟩⟩
```

Reply:
Halfway the proof, there seems to be a non-terminal `simp`.

Reply:
Yes originally I was having trouble with this one. When I did `squeeze_simp` its suggestion wouldn't work, but `simp` would. It seems to be working now, so I've fixed it.

Reply:
@CBirkbeck The above was also a suggestion to use `begin .. end` (not just the non-terminal simp).

Reply:
Ah sorry I missed that! Fixed now.